server/app/controllers/api/ProgramApplicationsApiController.java

```diff
@@ -226,10 +306,85 @@ private Instant parseParamDateToInstant(String paramName, String paramDate) {
+   *     specified
+   * @throws BadApiRequestException if pagination token and query parameter specify different values
+   */
+  private ImmutableList<LifecycleStage> resolveRevisionStateParam(
+      Optional<ApiPaginationTokenPayload> apiPaginationTokenPayload,
+      Optional<String> revisionStateParam) {
+    if (revisionStateParam.isEmpty()) {
+      return ImmutableList.of();
+    }
+
+    ImmutableList<LifecycleStage> tokenLifecycleStages =
+        apiPaginationTokenPayload
+            .map(
+                (token) ->
+                    extractLifecycleStagesFromPaginationToken(token, revisionStateParam.get()))
+            .orElse(ImmutableList.of());
+    ImmutableList<LifecycleStage> queryLifecycleStages = getLifecycleStages(revisionStateParam);
+
+    if (tokenLifecycleStages.equals(queryLifecycleStages)) {
+      return tokenLifecycleStages;
+    }
+
+    if (!tokenLifecycleStages.isEmpty() && !queryLifecycleStages.isEmpty()) {
+      throw new BadApiRequestException(
+          "Request parameters must match pagination token: " + revisionStateParam);
+    }
+
+    return !tokenLifecycleStages.isEmpty() ? tokenLifecycleStages : queryLifecycleStages;
+  }
+
+  /**
+   * Extracts the lifecycle stage from a pagination token's request specification.
+   *
+   * @param apiPaginationTokenPayload The pagination token containing previous request context
+   * @param paramName The parameter name to look for in the token's request spec
+   * @return Optional LifecycleStage if found in token, empty otherwise
+   */
+  private ImmutableList<LifecycleStage> extractLifecycleStagesFromPaginationToken(
+      ApiPaginationTokenPayload apiPaginationTokenPayload, String paramName) {
+    Map<String, String> requestSpec = apiPaginationTokenPayload.getRequestSpec();
+
+    if (!requestSpec.containsKey(paramName)) {
+      return ImmutableList.of();
+    }
+
+    String stagesString = requestSpec.get(REVISION_STATE_PARAM_NAME);
+    return Arrays.stream(stagesString.split(","))
+        .map(String::trim)
+        .map(LifecycleStage::valueOf)
+        .collect(ImmutableList.toImmutableList());
+  }
+
+  /**
+   * Converts a revision state parameter string to the corresponding LifecycleStage enum, or to the
+   * default values if none is provided,
+   *
+   * @param revisionStateParam Optional string representation of revision state
+   * @return Corresponding LifecycleStage enum values
+   * @throws BadApiRequestException if the parameter doesn't match any valid RevisionState
+   */
+  private ImmutableList<LifecycleStage> getLifecycleStages(Optional<String> revisionStateParam) {
+    // Default to all submitted application if no revision state is provided
+    if (revisionStateParam.isEmpty()) {
+      return ImmutableList.of(LifecycleStage.ACTIVE, LifecycleStage.OBSOLETE);
+    }
+    try {
+      RevisionState state = RevisionState.valueOf(revisionStateParam.get());
```

Comment:
This is case sensitive, since it should match the enum. We can make it case insensitive, though